modules/cas_cache/volume/vol_block_dev_top.c

```diff
@@ -247,22 +260,81 @@ static void _blockdev_handle_data(ocf_core_t core, struct bio *bio)
 	if (!io) {
 		printk(KERN_CRIT "Out of memory. Ending IO processing.\n");
 		cas_free_blk_data(data);
-		CAS_BIO_ENDIO(bio, CAS_BIO_BISIZE(bio), CAS_ERRNO_TO_BLK_STS(-ENOMEM));
-		return;
+		return -ENOMEM;
 	}
 
 	ret = ocf_io_set_data(io, data, 0);
 	if (ret < 0) {
 		ocf_io_put(io);
 		cas_free_blk_data(data);
-		CAS_BIO_ENDIO(bio, CAS_BIO_BISIZE(bio), CAS_ERRNO_TO_BLK_STS(-EINVAL));
-		return;
+		return -EINVAL;
 	}
 
-	ocf_io_set_cmpl(io, NULL, NULL, block_dev_complete_data);
-	data->start_time = cas_generic_start_io_acct(bio);
+	if (!master_ctx->data) {
+		atomic_set(&data->master_remaining, 1);
+		data->bio = master_ctx->bio;
+		data->master_size = master_ctx->master_size;
+		data->start_time = master_ctx->start_time;
+		master_ctx->data = data;
+	}
+
+	atomic_inc(&master_ctx->data->master_remaining);
+
+	ocf_io_set_cmpl(io, bio, master_ctx->data, block_dev_complete_data);
 
 	ocf_core_submit_io(io);
+
+	return 0;
+}
+
+static void _blockdev_handle_data(ocf_core_t core, struct bio *bio)
+{
+	const uint32_t max_io_sectors = (32*MiB) >> SECTOR_SHIFT;
+	const uint32_t align_sectors = (128*KiB) >> SECTOR_SHIFT;
+	struct bio *split;
+	uint32_t sectors, to_submit, unaligned;
+	int error;
+	struct blockdev_data_master_ctx master_ctx = {
+		.bio = bio,
+		.master_size = CAS_BIO_BISIZE(bio),
+	};
+
+	if (unlikely(CAS_BIO_BISIZE(bio) == 0)) {
+		CAS_PRINT_RL(KERN_ERR
+			"Not able to handle empty BIO, flags = "
+			CAS_BIO_OP_FLAGS_FORMAT "\n",  CAS_BIO_OP_FLAGS(bio));
+		error = -EINVAL;
+		goto err;
+	}
+
+	master_ctx.start_time = cas_generic_start_io_acct(bio);
+	for (sectors = bio_sectors(bio); sectors > 0;) {
+		if (sectors <= max_io_sectors) {
+			split = bio;
+			sectors = 0;
+		} else {
+			to_submit = max_io_sectors;
+			unaligned = CAS_BIO_BISECTOR(bio) % align_sectors;
+			if (unlikely(unaligned))
+				to_submit +=  unaligned - align_sectors;
+			split = cas_bio_split(bio, to_submit);
+			sectors -= to_submit;
+		}
+
+		error = _blockdev_handle_data_single(core, split, &master_ctx);
+		if (error)
+			goto err;
+	}
+
+	_block_dev_complete_data(master_ctx.data, 0);
+
+	return;
+
+err:
+	if (master_ctx.data)
+		_block_dev_complete_data(master_ctx.data, error);
+	else
+		CAS_BIO_ENDIO(bio, CAS_BIO_BISIZE(bio), CAS_ERRNO_TO_BLK_STS(error));
```

Comment:
My understanding is that this assumes split bio is not created. This might not be the case in case of error in the first call to _blockdev_handle_data_single.

Reply:
Good catch. Fixed.